<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.4" />
<title>cytograph.annotation.karyotyper API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/highlight.min.js" integrity="sha256-eOgo0OtLL4cdq7RdwRUiGKLX9XsIJ7nGhWEKbohmVAQ=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>cytograph.annotation.karyotyper</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from statsmodels.nonparametric.smoothers_lowess import lowess
from ..utils import div0
from ..module import requires, creates, Module
import numpy as np
import shoji


def windowed_mean(x: np.ndarray, n: int):
        if len(x) == 0:
                return x
        y = np.zeros_like(x)
        for ix in range(len(x)):
                i = min(ix, len(x) - n)
                w = x[i:i + n]
                y[ix] = np.mean(w)
        return y


class Karyotyper(Module):
        &#34;&#34;&#34;
        Estimate the karyotype of tumor cells using immune cells as internal reference
        &#34;&#34;&#34;
        def __init__(self, max_rv: int = 3, min_umis: int = 1, window_size: int = 20, smoothing_bandwidth: int = 0):
                &#34;&#34;&#34;
                Estimate the karyotype

                Args:
                        max_rv:              Max residual variance for selecting housekeeping genes
                        min_umis:            Minimum UMIs for selecting housekeeping genes
                        window_size:         Number of genes per window
                        smooting_bandwidth:  Optional bandwidth for smoothing

                Remarks:
                        This algorithm first identifies immune cells using the M-IMMUNE auto-annotation
                        (e.g. PTPRC expression) and computes a reference normal karyotype. It then
                        estimates the karyotype of each cluster and calls each as aneuploid or euploid.
                &#34;&#34;&#34;
                self.max_rv = max_rv
                self.min_umis = min_umis
                self.window_size = window_size
                self.smoothing_bandwidth = smoothing_bandwidth
                
                # The genome offsets of the starts of all the true chromosomes
                self.chr_starts = {
                        &#39;1&#39;: 0,
                        &#39;2&#39;: 248956422,
                        &#39;3&#39;: 491149951,
                        &#39;4&#39;: 689445510,
                        &#39;5&#39;: 879660065,
                        &#39;6&#39;: 1061198324,
                        &#39;7&#39;: 1232004303,
                        &#39;8&#39;: 1391350276,
                        &#39;9&#39;: 1536488912,
                        &#39;10&#39;: 1674883629,
                        &#39;11&#39;: 1808681051,
                        &#39;12&#39;: 1943767673,
                        &#39;13&#39;: 2077042982,
                        &#39;14&#39;: 2191407310,
                        &#39;15&#39;: 2298451028,
                        &#39;16&#39;: 2400442217,
                        &#39;17&#39;: 2490780562,
                        &#39;18&#39;: 2574038003,
                        &#39;19&#39;: 2654411288,
                        &#39;20&#39;: 2713028904,
                        &#39;21&#39;: 2777473071,
                        &#39;22&#39;: 2824183054,
                        &#39;X&#39;: 2875001522,
                        &#39;Y&#39;: 3031042417
                }

        @requires(&#34;PearsonResidualsVariance&#34;, &#34;float32&#34;, (&#34;genes&#34;,))
        @requires(&#34;GeneTotalUMIs&#34;, &#34;uint32&#34;, (&#34;genes&#34;,))
        @requires(&#34;Chromosome&#34;, &#34;string&#34;, (&#34;genes&#34;,))
        @requires(&#34;Start&#34;, &#34;int64&#34;, (&#34;genes&#34;,))
        @requires(&#34;ClusterID&#34;, &#34;uint32&#34;, (&#34;clusters&#34;,))
        @requires(&#34;MeanExpression&#34;, &#34;float64&#34;, (&#34;clusters&#34;, &#34;genes&#34;))
        @requires(&#34;AnnotationName&#34;, &#34;string&#34;, (&#34;annotations&#34;,))
        @requires(&#34;AnnotationPosterior&#34;, &#34;float32&#34;, (&#34;clusters&#34;, &#34;annotations&#34;))
        @creates(&#34;Aneuploid&#34;, &#34;bool&#34;, (&#34;clusters&#34;,))
        @creates(&#34;KaryotypePloidy&#34;, &#34;float32&#34;, (&#34;clusters&#34;, &#34;karyotype&#34;,))
        @creates(&#34;KaryotypePosition&#34;, &#34;float32&#34;, (&#34;karyotype&#34;,))
        @creates(&#34;KaryotypeChromosome&#34;, &#34;string&#34;, (&#34;karyotype&#34;,))
        @creates(&#34;ChromosomeStart&#34;, &#34;uint32&#34;, (&#34;chromosomes&#34;,))
        @creates(&#34;ChromosomeLength&#34;, &#34;uint32&#34;, (&#34;chromosomes&#34;,))
        @creates(&#34;ChromosomeName&#34;, &#34;string&#34;, (&#34;chromosomes&#34;,))
        def fit(self, ws: shoji.WorkspaceManager, save: bool = False):
                # Identify housekeeping genes
                self.housekeeping = (self.PearsonResidualsVariance[:] &lt; self.max_rv) &amp; ((self.GeneTotalUMIs[:] / ws.cells.length) &gt; self.min_umis)

                # Order by genomic position
                chrs = self.Chromosome[self.housekeeping]
                starts = self.Start[self.housekeeping]
                for chrom in self.chr_starts.keys():
                        starts[chrs == chrom] += self.chr_starts[chrom]
                self.ordering = np.argsort(starts)
                self.chromosomes = chrs[self.ordering]
                self.starts = starts[self.ordering]

                # Load the aggregated expression data
                y_sample = self.MeanExpression[:]
                totals = y_sample.sum(axis=1)
                y_sample = (y_sample.T / totals * np.median(totals)).T
                y_sample = y_sample[:, self.housekeeping]
                self.y_sample = y_sample[:, self.ordering]
                self.y_sample_mean = self.y_sample.mean(axis=1)

                # Load the immune cell clusters
                assert &#34;M-IMMUNE&#34; in self.AnnotationName[:], &#34;M-IMMUNE auto-annotation is required for Karyotyper&#34;
                self.immune = (self.AnnotationPosterior[:, self.AnnotationName[:] == &#34;M-IMMUNE&#34;] &gt; 0.9)[:, 0]
                y_ref = self.MeanExpression[self.immune, :]
                totals = y_ref.sum(axis=1)
                y_ref = (y_ref.T / totals * np.median(totals)).T
                y_ref = y_ref[:, self.housekeeping]
                self.y_ref = y_ref[:, self.ordering].mean(axis=0)
                self.y_ref_mean = self.y_ref.mean()

                # Bin locally along each chromosome
                for chrom in self.chr_starts.keys():
                        selected = (self.chromosomes == chrom)
                        if selected.sum() == 0:
                                continue
                        self.y_ref[selected] = windowed_mean(self.y_ref[selected], self.window_size)
                        for i in range(self.y_sample.shape[0]):
                                self.y_sample[i, selected] = windowed_mean(self.y_sample[i, selected], self.window_size)
                
                # Center using the reference residuals
                self.y_ratio = (div0(self.y_sample, self.y_ref).T * self.y_ref_mean / self.y_sample_mean).T

                if self.smoothing_bandwidth &gt; 0:
                        # Loess smoothing along each chromosome
                        self.y_ratio_smooth = np.copy(self.y_ratio)
                        for chrom in self.chr_starts.keys():
                                selected = (self.chromosomes == chrom)
                                for i in range(self.y_ratio_smooth.shape[0]):
                                        self.y_ratio_smooth[i, selected] = lowess(self.y_ratio_smooth[i, selected], self.starts[selected], frac=min(0.5, self.smoothing_bandwidth / selected.sum()), return_sorted=False)
                else:
                        self.y_ratio_smooth = None

                if &#34;karyotype&#34; in ws:
                        del ws.KaryotypePloidy
                        del ws.KaryotypePosition
                        del ws.KaryotypeChromosome
                        del ws.Aneuploid
                        del ws.karyotype

                ws.karyotype = shoji.Dimension(shape=self.y_ratio.shape[1])
                ws.chromosomes = shoji.Dimension(shape=len(self.chr_starts))

                sig_std = np.percentile(np.std(self.y_ratio[self.immune], axis=1), 95)
                aneuploid = np.std(self.y_ratio, axis=1)[self.ClusterID[:]] &gt; sig_std

                chr_starts = np.array(list(self.chr_starts.values()))
                chr_lengths = np.diff(np.append(chr_starts, int(3.04e9)))
                chr_names = np.array(list(self.chr_starts.keys()))

                return aneuploid, 2 * self.y_ratio, self.starts, self.chromosomes, chr_starts, chr_lengths, chr_names</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="cytograph.annotation.karyotyper.windowed_mean"><code class="name flex">
<span>def <span class="ident">windowed_mean</span></span>(<span>x: numpy.ndarray, n: int)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def windowed_mean(x: np.ndarray, n: int):
        if len(x) == 0:
                return x
        y = np.zeros_like(x)
        for ix in range(len(x)):
                i = min(ix, len(x) - n)
                w = x[i:i + n]
                y[ix] = np.mean(w)
        return y</code></pre>
</details>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="cytograph.annotation.karyotyper.Karyotyper"><code class="flex name class">
<span>class <span class="ident">Karyotyper</span></span>
<span>(</span><span>max_rv: int = 3, min_umis: int = 1, window_size: int = 20, smoothing_bandwidth: int = 0)</span>
</code></dt>
<dd>
<div class="desc"><p>Estimate the karyotype of tumor cells using immune cells as internal reference</p>
<p>Estimate the karyotype</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>max_rv</code></strong></dt>
<dd>
<pre><code>     Max residual variance for selecting housekeeping genes
</code></pre>
</dd>
<dt><strong><code>min_umis</code></strong></dt>
<dd>
<pre><code>   Minimum UMIs for selecting housekeeping genes
</code></pre>
</dd>
<dt><strong><code>window_size</code></strong></dt>
<dd>
<pre><code>Number of genes per window
</code></pre>
</dd>
<dt><strong><code>smooting_bandwidth</code></strong></dt>
<dd>Optional bandwidth for smoothing</dd>
</dl>
<h2 id="remarks">Remarks</h2>
<p>This algorithm first identifies immune cells using the M-IMMUNE auto-annotation
(e.g. PTPRC expression) and computes a reference normal karyotype. It then
estimates the karyotype of each cluster and calls each as aneuploid or euploid.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class Karyotyper(Module):
        &#34;&#34;&#34;
        Estimate the karyotype of tumor cells using immune cells as internal reference
        &#34;&#34;&#34;
        def __init__(self, max_rv: int = 3, min_umis: int = 1, window_size: int = 20, smoothing_bandwidth: int = 0):
                &#34;&#34;&#34;
                Estimate the karyotype

                Args:
                        max_rv:              Max residual variance for selecting housekeeping genes
                        min_umis:            Minimum UMIs for selecting housekeeping genes
                        window_size:         Number of genes per window
                        smooting_bandwidth:  Optional bandwidth for smoothing

                Remarks:
                        This algorithm first identifies immune cells using the M-IMMUNE auto-annotation
                        (e.g. PTPRC expression) and computes a reference normal karyotype. It then
                        estimates the karyotype of each cluster and calls each as aneuploid or euploid.
                &#34;&#34;&#34;
                self.max_rv = max_rv
                self.min_umis = min_umis
                self.window_size = window_size
                self.smoothing_bandwidth = smoothing_bandwidth
                
                # The genome offsets of the starts of all the true chromosomes
                self.chr_starts = {
                        &#39;1&#39;: 0,
                        &#39;2&#39;: 248956422,
                        &#39;3&#39;: 491149951,
                        &#39;4&#39;: 689445510,
                        &#39;5&#39;: 879660065,
                        &#39;6&#39;: 1061198324,
                        &#39;7&#39;: 1232004303,
                        &#39;8&#39;: 1391350276,
                        &#39;9&#39;: 1536488912,
                        &#39;10&#39;: 1674883629,
                        &#39;11&#39;: 1808681051,
                        &#39;12&#39;: 1943767673,
                        &#39;13&#39;: 2077042982,
                        &#39;14&#39;: 2191407310,
                        &#39;15&#39;: 2298451028,
                        &#39;16&#39;: 2400442217,
                        &#39;17&#39;: 2490780562,
                        &#39;18&#39;: 2574038003,
                        &#39;19&#39;: 2654411288,
                        &#39;20&#39;: 2713028904,
                        &#39;21&#39;: 2777473071,
                        &#39;22&#39;: 2824183054,
                        &#39;X&#39;: 2875001522,
                        &#39;Y&#39;: 3031042417
                }

        @requires(&#34;PearsonResidualsVariance&#34;, &#34;float32&#34;, (&#34;genes&#34;,))
        @requires(&#34;GeneTotalUMIs&#34;, &#34;uint32&#34;, (&#34;genes&#34;,))
        @requires(&#34;Chromosome&#34;, &#34;string&#34;, (&#34;genes&#34;,))
        @requires(&#34;Start&#34;, &#34;int64&#34;, (&#34;genes&#34;,))
        @requires(&#34;ClusterID&#34;, &#34;uint32&#34;, (&#34;clusters&#34;,))
        @requires(&#34;MeanExpression&#34;, &#34;float64&#34;, (&#34;clusters&#34;, &#34;genes&#34;))
        @requires(&#34;AnnotationName&#34;, &#34;string&#34;, (&#34;annotations&#34;,))
        @requires(&#34;AnnotationPosterior&#34;, &#34;float32&#34;, (&#34;clusters&#34;, &#34;annotations&#34;))
        @creates(&#34;Aneuploid&#34;, &#34;bool&#34;, (&#34;clusters&#34;,))
        @creates(&#34;KaryotypePloidy&#34;, &#34;float32&#34;, (&#34;clusters&#34;, &#34;karyotype&#34;,))
        @creates(&#34;KaryotypePosition&#34;, &#34;float32&#34;, (&#34;karyotype&#34;,))
        @creates(&#34;KaryotypeChromosome&#34;, &#34;string&#34;, (&#34;karyotype&#34;,))
        @creates(&#34;ChromosomeStart&#34;, &#34;uint32&#34;, (&#34;chromosomes&#34;,))
        @creates(&#34;ChromosomeLength&#34;, &#34;uint32&#34;, (&#34;chromosomes&#34;,))
        @creates(&#34;ChromosomeName&#34;, &#34;string&#34;, (&#34;chromosomes&#34;,))
        def fit(self, ws: shoji.WorkspaceManager, save: bool = False):
                # Identify housekeeping genes
                self.housekeeping = (self.PearsonResidualsVariance[:] &lt; self.max_rv) &amp; ((self.GeneTotalUMIs[:] / ws.cells.length) &gt; self.min_umis)

                # Order by genomic position
                chrs = self.Chromosome[self.housekeeping]
                starts = self.Start[self.housekeeping]
                for chrom in self.chr_starts.keys():
                        starts[chrs == chrom] += self.chr_starts[chrom]
                self.ordering = np.argsort(starts)
                self.chromosomes = chrs[self.ordering]
                self.starts = starts[self.ordering]

                # Load the aggregated expression data
                y_sample = self.MeanExpression[:]
                totals = y_sample.sum(axis=1)
                y_sample = (y_sample.T / totals * np.median(totals)).T
                y_sample = y_sample[:, self.housekeeping]
                self.y_sample = y_sample[:, self.ordering]
                self.y_sample_mean = self.y_sample.mean(axis=1)

                # Load the immune cell clusters
                assert &#34;M-IMMUNE&#34; in self.AnnotationName[:], &#34;M-IMMUNE auto-annotation is required for Karyotyper&#34;
                self.immune = (self.AnnotationPosterior[:, self.AnnotationName[:] == &#34;M-IMMUNE&#34;] &gt; 0.9)[:, 0]
                y_ref = self.MeanExpression[self.immune, :]
                totals = y_ref.sum(axis=1)
                y_ref = (y_ref.T / totals * np.median(totals)).T
                y_ref = y_ref[:, self.housekeeping]
                self.y_ref = y_ref[:, self.ordering].mean(axis=0)
                self.y_ref_mean = self.y_ref.mean()

                # Bin locally along each chromosome
                for chrom in self.chr_starts.keys():
                        selected = (self.chromosomes == chrom)
                        if selected.sum() == 0:
                                continue
                        self.y_ref[selected] = windowed_mean(self.y_ref[selected], self.window_size)
                        for i in range(self.y_sample.shape[0]):
                                self.y_sample[i, selected] = windowed_mean(self.y_sample[i, selected], self.window_size)
                
                # Center using the reference residuals
                self.y_ratio = (div0(self.y_sample, self.y_ref).T * self.y_ref_mean / self.y_sample_mean).T

                if self.smoothing_bandwidth &gt; 0:
                        # Loess smoothing along each chromosome
                        self.y_ratio_smooth = np.copy(self.y_ratio)
                        for chrom in self.chr_starts.keys():
                                selected = (self.chromosomes == chrom)
                                for i in range(self.y_ratio_smooth.shape[0]):
                                        self.y_ratio_smooth[i, selected] = lowess(self.y_ratio_smooth[i, selected], self.starts[selected], frac=min(0.5, self.smoothing_bandwidth / selected.sum()), return_sorted=False)
                else:
                        self.y_ratio_smooth = None

                if &#34;karyotype&#34; in ws:
                        del ws.KaryotypePloidy
                        del ws.KaryotypePosition
                        del ws.KaryotypeChromosome
                        del ws.Aneuploid
                        del ws.karyotype

                ws.karyotype = shoji.Dimension(shape=self.y_ratio.shape[1])
                ws.chromosomes = shoji.Dimension(shape=len(self.chr_starts))

                sig_std = np.percentile(np.std(self.y_ratio[self.immune], axis=1), 95)
                aneuploid = np.std(self.y_ratio, axis=1)[self.ClusterID[:]] &gt; sig_std

                chr_starts = np.array(list(self.chr_starts.values()))
                chr_lengths = np.diff(np.append(chr_starts, int(3.04e9)))
                chr_names = np.array(list(self.chr_starts.keys()))

                return aneuploid, 2 * self.y_ratio, self.starts, self.chromosomes, chr_starts, chr_lengths, chr_names</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="cytograph.module.Module" href="../module.html#cytograph.module.Module">Module</a></li>
</ul>
<h3>Methods</h3>
<dl>
<dt id="cytograph.annotation.karyotyper.Karyotyper.fit"><code class="name flex">
<span>def <span class="ident">fit</span></span>(<span>self, ws: shoji.workspace.WorkspaceManager, save: bool = False)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@requires(&#34;PearsonResidualsVariance&#34;, &#34;float32&#34;, (&#34;genes&#34;,))
@requires(&#34;GeneTotalUMIs&#34;, &#34;uint32&#34;, (&#34;genes&#34;,))
@requires(&#34;Chromosome&#34;, &#34;string&#34;, (&#34;genes&#34;,))
@requires(&#34;Start&#34;, &#34;int64&#34;, (&#34;genes&#34;,))
@requires(&#34;ClusterID&#34;, &#34;uint32&#34;, (&#34;clusters&#34;,))
@requires(&#34;MeanExpression&#34;, &#34;float64&#34;, (&#34;clusters&#34;, &#34;genes&#34;))
@requires(&#34;AnnotationName&#34;, &#34;string&#34;, (&#34;annotations&#34;,))
@requires(&#34;AnnotationPosterior&#34;, &#34;float32&#34;, (&#34;clusters&#34;, &#34;annotations&#34;))
@creates(&#34;Aneuploid&#34;, &#34;bool&#34;, (&#34;clusters&#34;,))
@creates(&#34;KaryotypePloidy&#34;, &#34;float32&#34;, (&#34;clusters&#34;, &#34;karyotype&#34;,))
@creates(&#34;KaryotypePosition&#34;, &#34;float32&#34;, (&#34;karyotype&#34;,))
@creates(&#34;KaryotypeChromosome&#34;, &#34;string&#34;, (&#34;karyotype&#34;,))
@creates(&#34;ChromosomeStart&#34;, &#34;uint32&#34;, (&#34;chromosomes&#34;,))
@creates(&#34;ChromosomeLength&#34;, &#34;uint32&#34;, (&#34;chromosomes&#34;,))
@creates(&#34;ChromosomeName&#34;, &#34;string&#34;, (&#34;chromosomes&#34;,))
def fit(self, ws: shoji.WorkspaceManager, save: bool = False):
        # Identify housekeeping genes
        self.housekeeping = (self.PearsonResidualsVariance[:] &lt; self.max_rv) &amp; ((self.GeneTotalUMIs[:] / ws.cells.length) &gt; self.min_umis)

        # Order by genomic position
        chrs = self.Chromosome[self.housekeeping]
        starts = self.Start[self.housekeeping]
        for chrom in self.chr_starts.keys():
                starts[chrs == chrom] += self.chr_starts[chrom]
        self.ordering = np.argsort(starts)
        self.chromosomes = chrs[self.ordering]
        self.starts = starts[self.ordering]

        # Load the aggregated expression data
        y_sample = self.MeanExpression[:]
        totals = y_sample.sum(axis=1)
        y_sample = (y_sample.T / totals * np.median(totals)).T
        y_sample = y_sample[:, self.housekeeping]
        self.y_sample = y_sample[:, self.ordering]
        self.y_sample_mean = self.y_sample.mean(axis=1)

        # Load the immune cell clusters
        assert &#34;M-IMMUNE&#34; in self.AnnotationName[:], &#34;M-IMMUNE auto-annotation is required for Karyotyper&#34;
        self.immune = (self.AnnotationPosterior[:, self.AnnotationName[:] == &#34;M-IMMUNE&#34;] &gt; 0.9)[:, 0]
        y_ref = self.MeanExpression[self.immune, :]
        totals = y_ref.sum(axis=1)
        y_ref = (y_ref.T / totals * np.median(totals)).T
        y_ref = y_ref[:, self.housekeeping]
        self.y_ref = y_ref[:, self.ordering].mean(axis=0)
        self.y_ref_mean = self.y_ref.mean()

        # Bin locally along each chromosome
        for chrom in self.chr_starts.keys():
                selected = (self.chromosomes == chrom)
                if selected.sum() == 0:
                        continue
                self.y_ref[selected] = windowed_mean(self.y_ref[selected], self.window_size)
                for i in range(self.y_sample.shape[0]):
                        self.y_sample[i, selected] = windowed_mean(self.y_sample[i, selected], self.window_size)
        
        # Center using the reference residuals
        self.y_ratio = (div0(self.y_sample, self.y_ref).T * self.y_ref_mean / self.y_sample_mean).T

        if self.smoothing_bandwidth &gt; 0:
                # Loess smoothing along each chromosome
                self.y_ratio_smooth = np.copy(self.y_ratio)
                for chrom in self.chr_starts.keys():
                        selected = (self.chromosomes == chrom)
                        for i in range(self.y_ratio_smooth.shape[0]):
                                self.y_ratio_smooth[i, selected] = lowess(self.y_ratio_smooth[i, selected], self.starts[selected], frac=min(0.5, self.smoothing_bandwidth / selected.sum()), return_sorted=False)
        else:
                self.y_ratio_smooth = None

        if &#34;karyotype&#34; in ws:
                del ws.KaryotypePloidy
                del ws.KaryotypePosition
                del ws.KaryotypeChromosome
                del ws.Aneuploid
                del ws.karyotype

        ws.karyotype = shoji.Dimension(shape=self.y_ratio.shape[1])
        ws.chromosomes = shoji.Dimension(shape=len(self.chr_starts))

        sig_std = np.percentile(np.std(self.y_ratio[self.immune], axis=1), 95)
        aneuploid = np.std(self.y_ratio, axis=1)[self.ClusterID[:]] &gt; sig_std

        chr_starts = np.array(list(self.chr_starts.values()))
        chr_lengths = np.diff(np.append(chr_starts, int(3.04e9)))
        chr_names = np.array(list(self.chr_starts.keys()))

        return aneuploid, 2 * self.y_ratio, self.starts, self.chromosomes, chr_starts, chr_lengths, chr_names</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="cytograph.annotation" href="index.html">cytograph.annotation</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="cytograph.annotation.karyotyper.windowed_mean" href="#cytograph.annotation.karyotyper.windowed_mean">windowed_mean</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="cytograph.annotation.karyotyper.Karyotyper" href="#cytograph.annotation.karyotyper.Karyotyper">Karyotyper</a></code></h4>
<ul class="">
<li><code><a title="cytograph.annotation.karyotyper.Karyotyper.fit" href="#cytograph.annotation.karyotyper.Karyotyper.fit">fit</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.4</a>.</p>
</footer>
</body>
</html>