<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.4" />
<title>cytograph.plotting.qc_plots API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/highlight.min.js" integrity="sha256-eOgo0OtLL4cdq7RdwRUiGKLX9XsIJ7nGhWEKbohmVAQ=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>cytograph.plotting.qc_plots</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import matplotlib.pyplot as plt
import numpy as np
import loompy
from .colors import colorize
from cytograph.enrichment import FeatureSelectionByVariance
from cytograph.embedding import art_of_tsne
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE


def all_QC_plots(ds: loompy.LoomConnection = None, out_file: str = &#34;Tmp.png&#34;):
        f, ax = plt.subplots(4, 2, figsize=(2 * 8, 4 * 8))
        
        plot_title = &#34;Mitochondrial gene expression ratio distribution&#34;
        dist_attr(ax[0, 0], ds, attr = &#34;MT_ratio&#34;, plot_title= plot_title, line=0.05)
        plot_title = &#34;Unspliced reads ratio distribution&#34;
        dist_attr(ax[0, 1], ds, attr = &#34;unspliced_ratio&#34;, plot_title= plot_title, line=0.5)
        attrs_on_TSNE(ax[1:4, 0:2], ds, attrs = [&#34;DoubletFinderFlag&#34;, &#34;DoubletFinderScore&#34;, &#34;TotalUMI&#34;, &#34;NGenes&#34;, &#34;unspliced_ratio&#34;, &#34;MT_ratio&#34;], plot_title = [&#34;Doublets Flag&#34;, &#34;Doublets Score&#34;, &#34;UMI counts&#34;, &#34;Number of genes per cell&#34;, &#34;Unspliced UMI ratio&#34;, &#34;Mitochondrial genes ratio&#34;])
        
        f.savefig(out_file, dpi=144)
        plt.close(f)


def dist_attr(ax: plt.axes = None, ds: loompy.LoomConnection = None, out_file: str = None, attr: str = None, plot_title: str = None, line: float = None):

        if ax is None:
                fig, ax = plt.subplots(figsize=(6, 4))
        if attr in ds.ca:
                ax.hist(ds.ca[attr], bins = 100)
                ax.set_xlim(np.amin(ds.ca[attr]), np.amax(ds.ca[attr]))
                if line is not None:
                        ax.axvline(x=line, c=&#39;r&#39;)
                if plot_title is not None:
                        ax.set_title(plot_title)
                if out_file is not None:
                        fig.savefig(out_file, dpi=144)
                return(ax)


def attrs_on_TSNE(ax: plt.axes = None, ds: loompy.LoomConnection = None, out_file: str = None, attrs: list = None, plot_title: list = None):
        n_attr = len(attrs)
        n_cols = 2
        n_rows = np.ceil(n_attr / n_cols)
        n_rows = n_rows.astype(int)
        if ax is None and n_attr &gt; 1:
                fig, ax = plt.subplots(n_rows, n_cols, figsize=(8 * n_cols, 8 * n_rows))
        elif ax is None and n_attr == 1:
                fig, ax = plt.subplots(figsize=(12, 12))
        n_cols_plot = 0
        n_rows_plot = 0
        if &#39;TSNE&#39; in ds.ca:
                xy = ds.ca.TSNE
        elif &#39;HPF&#39; in ds.ca:
                xy = art_of_tsne(ds.ca.HPF)
                ds.ca.TSNE = xy
        elif &#39;PCA&#39; in ds.ca:
                angle = 0.5
                perplexity = 30
                verbose = False
                xy = TSNE(angle=angle, perplexity=perplexity, verbose=verbose).fit_transform(ds.ca.PCA)
                ds.ca.TSNE = xy
        else:

                genes = FeatureSelectionByVariance(2000).fit(ds)
                data = ds[:, :]
                f = np.divide(data.sum(axis=0), 10e6)
                norm_data = np.divide(data, f)
                norm_data = np.log(norm_data + 1)
                ds.ca.PCA = PCA(n_components=50).fit_transform(norm_data[genes, :].T)
                angle = 0.5
                perplexity = 30
                verbose = False
                xy = TSNE(angle=angle, perplexity=perplexity, verbose=verbose).fit_transform(ds.ca.PCA)
                ds.ca.TSNE = xy
        
        for n, attr in enumerate(attrs):
                # ax.append( plt.subplot(n_rows,n_cols,n+1))
                if(n_attr == 1):
                        current_ax = ax
                if(n_attr == 2):
                        current_ax = ax[n_cols_plot]
                if(n_attr &gt; 2):
                        current_ax = ax[n_rows_plot, n_cols_plot]
                if attr in ds.ca:
                        current_ax.scatter(xy[:, 0], xy[:, 1], c=&#39;lightgrey&#39;, lw=0, s=10)
                        names, labels = np.unique(ds.ca[attr], return_inverse=True)
                        if (len(names) &gt; 20):
                                labels = ds.ca[attr]
                                if(np.max(ds.ca[attr]) &gt; 10000):
                                        labels = np.log(labels)
                                        if plot_title is not None:
                                                current_ax.set_title(plot_title[n] + &#34; (log transformed)&#34;)
                                elif plot_title is not None:
                                        current_ax.set_title(plot_title[n])
                                cm = plt.cm.get_cmap(&#39;jet&#39;)
                                cells = ds.ca[attr] &gt; 0
                                sc = current_ax.scatter(xy[:, 0][cells], xy[:, 1][cells], c=labels[cells], lw=0, s=10, cmap = cm)
                                plt.colorbar(sc, ax=current_ax)
                        else:
                                colors = colorize(names)
                                cells = np.random.permutation(labels.shape[0])
                                current_ax.scatter(xy[cells, 0], xy[cells, 1], c=colors[labels][cells], lw=0, s=10)

                                def h(c):
                                        return plt.Line2D([], [], color=c, ls=&#34;&#34;, marker=&#34;o&#34;)
                                current_ax.legend(handles=[h(colors[i]) for i in range(len(names))], labels=list(names), loc=&#39;lower left&#39;, markerscale=1, frameon=False, fontsize=10)
                                if plot_title is not None:
                                        current_ax.set_title(plot_title[n])
                else:
                        current_ax.scatter(xy[:, 0], xy[:, 1], c=&#39;lightgrey&#39;, lw=0, s=10)
                        if plot_title is not None:
                                current_ax.set_title(plot_title[n])
                if(n_cols_plot != n_cols - 1):
                        n_cols_plot = n_cols_plot + 1
                else:
                        n_cols_plot = 0
                        n_rows_plot = n_rows_plot + 1
        if out_file is not None:
                fig.savefig(out_file, dpi=144)
        return(ax)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="cytograph.plotting.qc_plots.all_QC_plots"><code class="name flex">
<span>def <span class="ident">all_QC_plots</span></span>(<span>ds: loompy.loompy.LoomConnection = None, out_file: str = 'Tmp.png')</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def all_QC_plots(ds: loompy.LoomConnection = None, out_file: str = &#34;Tmp.png&#34;):
        f, ax = plt.subplots(4, 2, figsize=(2 * 8, 4 * 8))
        
        plot_title = &#34;Mitochondrial gene expression ratio distribution&#34;
        dist_attr(ax[0, 0], ds, attr = &#34;MT_ratio&#34;, plot_title= plot_title, line=0.05)
        plot_title = &#34;Unspliced reads ratio distribution&#34;
        dist_attr(ax[0, 1], ds, attr = &#34;unspliced_ratio&#34;, plot_title= plot_title, line=0.5)
        attrs_on_TSNE(ax[1:4, 0:2], ds, attrs = [&#34;DoubletFinderFlag&#34;, &#34;DoubletFinderScore&#34;, &#34;TotalUMI&#34;, &#34;NGenes&#34;, &#34;unspliced_ratio&#34;, &#34;MT_ratio&#34;], plot_title = [&#34;Doublets Flag&#34;, &#34;Doublets Score&#34;, &#34;UMI counts&#34;, &#34;Number of genes per cell&#34;, &#34;Unspliced UMI ratio&#34;, &#34;Mitochondrial genes ratio&#34;])
        
        f.savefig(out_file, dpi=144)
        plt.close(f)</code></pre>
</details>
</dd>
<dt id="cytograph.plotting.qc_plots.attrs_on_TSNE"><code class="name flex">
<span>def <span class="ident">attrs_on_TSNE</span></span>(<span>ax: <function axes at 0x7ff1d8d708b0> = None, ds: loompy.loompy.LoomConnection = None, out_file: str = None, attrs: list = None, plot_title: list = None)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def attrs_on_TSNE(ax: plt.axes = None, ds: loompy.LoomConnection = None, out_file: str = None, attrs: list = None, plot_title: list = None):
        n_attr = len(attrs)
        n_cols = 2
        n_rows = np.ceil(n_attr / n_cols)
        n_rows = n_rows.astype(int)
        if ax is None and n_attr &gt; 1:
                fig, ax = plt.subplots(n_rows, n_cols, figsize=(8 * n_cols, 8 * n_rows))
        elif ax is None and n_attr == 1:
                fig, ax = plt.subplots(figsize=(12, 12))
        n_cols_plot = 0
        n_rows_plot = 0
        if &#39;TSNE&#39; in ds.ca:
                xy = ds.ca.TSNE
        elif &#39;HPF&#39; in ds.ca:
                xy = art_of_tsne(ds.ca.HPF)
                ds.ca.TSNE = xy
        elif &#39;PCA&#39; in ds.ca:
                angle = 0.5
                perplexity = 30
                verbose = False
                xy = TSNE(angle=angle, perplexity=perplexity, verbose=verbose).fit_transform(ds.ca.PCA)
                ds.ca.TSNE = xy
        else:

                genes = FeatureSelectionByVariance(2000).fit(ds)
                data = ds[:, :]
                f = np.divide(data.sum(axis=0), 10e6)
                norm_data = np.divide(data, f)
                norm_data = np.log(norm_data + 1)
                ds.ca.PCA = PCA(n_components=50).fit_transform(norm_data[genes, :].T)
                angle = 0.5
                perplexity = 30
                verbose = False
                xy = TSNE(angle=angle, perplexity=perplexity, verbose=verbose).fit_transform(ds.ca.PCA)
                ds.ca.TSNE = xy
        
        for n, attr in enumerate(attrs):
                # ax.append( plt.subplot(n_rows,n_cols,n+1))
                if(n_attr == 1):
                        current_ax = ax
                if(n_attr == 2):
                        current_ax = ax[n_cols_plot]
                if(n_attr &gt; 2):
                        current_ax = ax[n_rows_plot, n_cols_plot]
                if attr in ds.ca:
                        current_ax.scatter(xy[:, 0], xy[:, 1], c=&#39;lightgrey&#39;, lw=0, s=10)
                        names, labels = np.unique(ds.ca[attr], return_inverse=True)
                        if (len(names) &gt; 20):
                                labels = ds.ca[attr]
                                if(np.max(ds.ca[attr]) &gt; 10000):
                                        labels = np.log(labels)
                                        if plot_title is not None:
                                                current_ax.set_title(plot_title[n] + &#34; (log transformed)&#34;)
                                elif plot_title is not None:
                                        current_ax.set_title(plot_title[n])
                                cm = plt.cm.get_cmap(&#39;jet&#39;)
                                cells = ds.ca[attr] &gt; 0
                                sc = current_ax.scatter(xy[:, 0][cells], xy[:, 1][cells], c=labels[cells], lw=0, s=10, cmap = cm)
                                plt.colorbar(sc, ax=current_ax)
                        else:
                                colors = colorize(names)
                                cells = np.random.permutation(labels.shape[0])
                                current_ax.scatter(xy[cells, 0], xy[cells, 1], c=colors[labels][cells], lw=0, s=10)

                                def h(c):
                                        return plt.Line2D([], [], color=c, ls=&#34;&#34;, marker=&#34;o&#34;)
                                current_ax.legend(handles=[h(colors[i]) for i in range(len(names))], labels=list(names), loc=&#39;lower left&#39;, markerscale=1, frameon=False, fontsize=10)
                                if plot_title is not None:
                                        current_ax.set_title(plot_title[n])
                else:
                        current_ax.scatter(xy[:, 0], xy[:, 1], c=&#39;lightgrey&#39;, lw=0, s=10)
                        if plot_title is not None:
                                current_ax.set_title(plot_title[n])
                if(n_cols_plot != n_cols - 1):
                        n_cols_plot = n_cols_plot + 1
                else:
                        n_cols_plot = 0
                        n_rows_plot = n_rows_plot + 1
        if out_file is not None:
                fig.savefig(out_file, dpi=144)
        return(ax)</code></pre>
</details>
</dd>
<dt id="cytograph.plotting.qc_plots.dist_attr"><code class="name flex">
<span>def <span class="ident">dist_attr</span></span>(<span>ax: <function axes at 0x7ff1d8d708b0> = None, ds: loompy.loompy.LoomConnection = None, out_file: str = None, attr: str = None, plot_title: str = None, line: float = None)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def dist_attr(ax: plt.axes = None, ds: loompy.LoomConnection = None, out_file: str = None, attr: str = None, plot_title: str = None, line: float = None):

        if ax is None:
                fig, ax = plt.subplots(figsize=(6, 4))
        if attr in ds.ca:
                ax.hist(ds.ca[attr], bins = 100)
                ax.set_xlim(np.amin(ds.ca[attr]), np.amax(ds.ca[attr]))
                if line is not None:
                        ax.axvline(x=line, c=&#39;r&#39;)
                if plot_title is not None:
                        ax.set_title(plot_title)
                if out_file is not None:
                        fig.savefig(out_file, dpi=144)
                return(ax)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="cytograph.plotting" href="index.html">cytograph.plotting</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="cytograph.plotting.qc_plots.all_QC_plots" href="#cytograph.plotting.qc_plots.all_QC_plots">all_QC_plots</a></code></li>
<li><code><a title="cytograph.plotting.qc_plots.attrs_on_TSNE" href="#cytograph.plotting.qc_plots.attrs_on_TSNE">attrs_on_TSNE</a></code></li>
<li><code><a title="cytograph.plotting.qc_plots.dist_attr" href="#cytograph.plotting.qc_plots.dist_attr">dist_attr</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.4</a>.</p>
</footer>
</body>
</html>